test/infrastructure/docker/internal/controllers/dockermachine_controller.go

```diff
@@ -288,26 +288,34 @@ func (r *DockerMachineReconciler) reconcileNormal(ctx context.Context, cluster *
 
 	// if the machine isn't bootstrapped, only then run bootstrap scripts
 	if !dockerMachine.Spec.Bootstrapped {
-		bootstrapData, format, err := r.getBootstrapData(ctx, machine)
-		if err != nil {
-			log.Error(err, "failed to get bootstrap data")
-			return ctrl.Result{}, err
-		}
-
-		timeoutctx, cancel := context.WithTimeout(ctx, 3*time.Minute)
+		timeoutCtx, cancel := context.WithTimeout(ctx, 3*time.Minute)
 		defer cancel()
-		// Run the bootstrap script. Simulates cloud-init/Ignition.
-		if err := externalMachine.ExecBootstrap(timeoutctx, bootstrapData, format); err != nil {
-			conditions.MarkFalse(dockerMachine, infrav1.BootstrapExecSucceededCondition, infrav1.BootstrapFailedReason, clusterv1.ConditionSeverityWarning, "Repeating bootstrap")
-			return ctrl.Result{}, errors.Wrap(err, "failed to exec DockerMachine bootstrap")
-		}
+
 		// Check for bootstrap success
-		if err := externalMachine.CheckForBootstrapSuccess(timeoutctx); err != nil {
-			conditions.MarkFalse(dockerMachine, infrav1.BootstrapExecSucceededCondition, infrav1.BootstrapFailedReason, clusterv1.ConditionSeverityWarning, "Repeating bootstrap")
-			return ctrl.Result{}, errors.Wrap(err, "failed to check for existence of bootstrap success file at /run/cluster-api/bootstrap-success.complete")
-		}
+		// We have to check here to make this reentrant for cases where the bootstrap works
+		// but bootstrapped is never set on the object.
+		if err := externalMachine.CheckForBootstrapSuccess(timeoutCtx, false); err == nil {
+			dockerMachine.Spec.Bootstrapped = true
+		} else {
```

Comment:
We could move setting `dockerMachine.Spec.Bootstrapped = true` below the whole if and only have
```go
		if err := externalMachine.CheckForBootstrapSuccess(timeoutCtx, false); err != nil {
```
here, wdyt?

Reply:
Sgtm

Reply:
https://github.com/kubernetes-sigs/cluster-api/pull/7271